firewood/src/merkle/node.rs

```diff
@@ -402,6 +416,113 @@ impl<T> PartialEq for EncodedNode<T> {
     }
 }
 
+#[derive(Debug)]
+struct PathWithBitsPrefix(Path);
+
+impl<'de> Deserialize<'de> for PathWithBitsPrefix {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        // TODO danlaine: This is slightly abusing the serde API.
+        // We don't know the number of bytes in the tuple until we've read the first u64.
+        // We should implement a custom serializer/deserializer for EncodedNode.
+        deserializer.deserialize_tuple(usize::MAX, TupleVisitor)
+    }
+}
+
+struct TupleVisitor;
+
+impl<'de> Visitor<'de> for TupleVisitor {
+    type Value = PathWithBitsPrefix;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("tuple")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        let path_bits_len = seq
+            .next_element::<u64>()?
+            .ok_or(serde::de::Error::custom("missing length"))?;
+
+        if path_bits_len == 0 {
+            return Ok(PathWithBitsPrefix(Path(vec![])));
+        }
+
+        if path_bits_len % BITS_PER_NIBBLE != 0 {
+            return Err(serde::de::Error::custom(
+                "path length is not a multiple of 4",
+            ));
+        }
+
+        if path_bits_len > MAX_PATH_BIT_LEN {
+            return Err(serde::de::Error::custom(
+                "path bit length > {MAX_PATH_BIT_LEN}",
+            ));
+        }
+
+        // Add BITS_PER_NIBBLE so that if there are an odd number of nibbles,
+        // we need to read the final byte.
+        // e.g. if `path_bits_len` is 12 we need to read 2 bytes
+        // If we didn't add BITS_PER_NIBBLE `path_bytes_len` would be 1.
+        // Note `path_bytes_len` >= 1 because we assert `path_bits_len` > 0.
+        let path_bytes_len = (path_bits_len + BITS_PER_NIBBLE) / BITS_PER_BYTE;
+
+        let mut path = Vec::with_capacity(NIBBLES_PER_BYTE * path_bytes_len as usize);
+
+        for _ in 0..path_bytes_len {
+            let byte = seq
+                .next_element::<u8>()?
+                .ok_or(serde::de::Error::custom("not enough bytes"))?;
+
+            // Convert the byte to 2 nibbles
+            path.push(byte >> 4);
+            path.push(byte & 0b_0000_1111);
+        }
+
+        if path_bits_len % BITS_PER_BYTE != 0 {
+            // The last byte only contained one nibble.
+            let padding_nibble = path.pop().expect("path_bytes_len > 0");
+            if padding_nibble != 0 {
+                return Err(serde::de::Error::custom("padding nibble is not 0"));
+            }
+        }
```

Comment:
Why not do this check outside the loop and determine the range based on the check?